core/src/main/java/hudson/model/ItemGroupMixIn.java

```diff
@@ -329,6 +338,38 @@ public synchronized TopLevelItem createProject( TopLevelItemDescriptor type, Str
     }
 
     /**
+     * Populate a {@link Categories} object from a specific {@link ItemGroup}.
+     *
+     * @return A object that represents a set of {@link Category}
+     */
+    public static Categories getCategories(Authentication a, ItemGroup c) {
+        Categories categories = new Categories();
+        int weight = ItemCategory.MIN_WEIGHT;
+        for (TopLevelItemDescriptor descriptor : DescriptorVisibilityFilter.apply(c, Items.all(a, c))) {
+            ItemCategory ic = ItemCategoryConfigurator.getCategory(descriptor);
+            Map<String, Serializable> metadata = new HashMap<String, Serializable>();
+
+            // Information about Item.
+            metadata.put("class", descriptor.getId());
+            metadata.put("weight", ++weight);
+            metadata.put("displayName", descriptor.getDisplayName());
+            metadata.put("description", descriptor.getDescription());
+            metadata.put("iconFilePathPattern", descriptor.getIconFilePathPattern());
+
+            Category category = categories.getItem(ic.getId());
+            if (category != null) {
+                category.getItems().add(metadata);
```

Comment:
Similar to @oleg-nenashev’s old note in `Category`, it is not great style to be creating a mutable object, when there is no real reason to mutate it outside this method.

Again if the class is restricted, then it does not much matter since a local “find usages” can show where all the mutation is, and we can see that it is safe.